node/src/certificate_store.rs

```diff
@@ -0,0 +1,238 @@
+    /// Holds the certificates by their digest id
+    certificates_by_id: DBMap<CertificateDigest, Certificate>,
+    /// A secondary index that keeps the certificate digest ids
+    /// by the certificate rounds. That helps us to perform
+    /// range requests based on rounds. We avoid storing again the
+    /// certificate here to not waste space. To dereference we use
+    /// the certificates_by_id storage.
+    certificate_ids_by_round: DBMap<(Round, CertificateDigest), u8>,
+    /// Senders to notify for a write that happened for
+    /// the specified certificate digest id
+    obligations: Arc<RwLock<HashMap<CertificateDigest, VecDeque<oneshot::Sender<Certificate>>>>>,
+}
+
+impl CertificateStore {
+    pub fn new(
+        certificates_by_id: DBMap<CertificateDigest, Certificate>,
+        certificate_ids_by_round: DBMap<(Round, CertificateDigest), u8>,
+    ) -> CertificateStore {
+        Self {
+            certificates_by_id,
+            certificate_ids_by_round,
+            obligations: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Inserts a certificate to the store
+    pub fn insert(&self, certificate: Certificate) -> StoreResult<()> {
+        let mut batch = self.certificates_by_id.batch();
+
+        let id = certificate.digest();
+        let round = certificate.round();
+
+        // write the certificate by its id
+        batch = batch.insert_batch(&self.certificates_by_id, vec![(id, certificate.clone())])?;
+
+        // write the certificate id by its round
+        let key = (round, id);
+        let value = 0;
+
+        batch = batch.insert_batch(&self.certificate_ids_by_round, vec![(key, value)])?;
+
+        // execute the batch (atomically) and return the result
+        let result = batch.write();
+
+        if result.is_ok() {
+            {
+                // now notify the obligations
+                let mut lock = self.obligations.write().unwrap();
+
+                if let Some(mut senders) = lock.remove(&id) {
+                    while let Some(s) = senders.pop_front() {
+                        let _ = s.send(certificate.clone());
+                    }
+                }
+            }
+        }
+
+        result
+    }
+
+    /// Retrieves a certificate from the store. If not found
+    /// then None is returned as result.
+    pub fn read(&self, id: CertificateDigest) -> StoreResult<Option<Certificate>> {
+        self.certificates_by_id.get(&id)
+    }
+
+    /// Retrieves multiple certificates by their provided ids. The results
+    /// are returned in the same sequence as the provided keys.
+    pub fn multi_read(
+        &self,
+        ids: impl IntoIterator<Item = CertificateDigest>,
+    ) -> StoreResult<Vec<Option<Certificate>>> {
+        self.certificates_by_id.multi_get(ids)
+    }
+
+    /// Waits to get notified until the requested certificate becomes available
+    pub async fn notify_read(&self, id: CertificateDigest) -> StoreResult<Option<Certificate>> {
```

Comment:
To completely ditch the `certificate_store` we need the pub/sub system the `Store` offers. Here I preferred I followed a different approach than Store - meaning not use a tokio task to handle the commands. Let me know yourr thoughts @huitseeker